File: server/sonar-qa-util/src/main/java/org/sonarqube/qa/util/pageobjects/settings/SettingsPage.java
/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarqube.qa.util.pageobjects.settings;

import com.codeborne.selenide.Condition;
import com.codeborne.selenide.Selenide;
import com.codeborne.selenide.SelenideElement;
import org.openqa.selenium.By;
import org.openqa.selenium.Keys;

import static com.codeborne.selenide.Selectors.byText;

public class SettingsPage {

  public SettingsPage() {
    Selenide.$("#settings-page").shouldBe(Condition.visible);
  }

  public SettingsPage assertMenuContains(String categoryName) {
    Selenide.$(".side-tabs-menu").$(By.linkText(categoryName)).shouldBe(Condition.visible);
    return this;
  }

  public SettingsPage assertSettingDisplayed(String settingKey) {
    Selenide.$(".settings-definition[data-key='" + settingKey + "']").shouldBe(Condition.visible);
    return this;
  }

  public SettingsPage assertSettingNotDisplayed(String settingKey) {
    Selenide.$(".settings-definition[data-key='" + settingKey + "']").shouldNotBe(Condition.visible);
    return this;
  }

  public SettingsPage openCategory(String categoryName) {
    Selenide.$(".side-tabs-menu").$(By.linkText(categoryName)).click();
    return this;
  }

  public SettingsPage assertStringSettingValue(String settingKey, String value) {
    Selenide.$("input[name=\"settings[" + settingKey + "]\"]").shouldHave(Condition.exactValue(value));
    return this;
  }

  public SettingsPage assertSettingValueIsNotedAsDefault(String settingKey) {
    Selenide.$("[data-key=\"" + settingKey + "\"]").find(".spacer-top.note")
      .shouldBe(Condition.text("(default)"));
    return this;
  }

  public SettingsPage assertBooleanSettingValue(String settingKey, boolean value) {
    SelenideElement toggle = Selenide.$("button[name=\"settings[" + settingKey + "]\"]");
    if (value) {
      toggle.shouldHave(Condition.cssClass("boolean-toggle-on"));
    } else {
      toggle.shouldNotHave(Condition.cssClass("boolean-toggle-on"));
    }
    return this;
  }

  public SettingsPage assertSettingValueCanBeSaved(String settingKey) {
    Selenide.$("[data-key=\"" + settingKey + "\"]").find(byText("Save"))
      .should(Condition.exist)
      .shouldNotBe(Condition.attribute("disabled"));
    return this;
  }

  public SettingsPage assertSettingValueCannotBeSaved(String settingKey) {
    Selenide.$("[data-key=\"" + settingKey + "\"]").find(byText("Save"))
      .should(Condition.exist)
      .shouldBe(Condition.attribute("disabled"));
    return this;
  }

  public SettingsPage assertSettingValueCanBeReset(String settingKey) {
    Selenide.$("[data-key=\"" + settingKey + "\"]").find(byText("Reset"))
      .should(Condition.exist);
    return this;
  }

  public SettingsPage assertSettingValueCanBeCanceled(String settingKey) {
    Selenide.$("[data-key=\"" + settingKey + "\"]").find(byText("Cancel"))
      .should(Condition.exist);
    return this;
  }

  public SettingsPage assertInputCount(String settingKey, int count) {
    Selenide.$("[data-key=\"" + settingKey + "\"]").findAll("input").shouldHaveSize(count);
    return this;
  }

  public SettingsPage changeSettingValue(String settingKey, String value) {
    Selenide.$("[data-key=\"" + settingKey + "\"]").find("input").val(value);
    return this;
  }

  public SettingsPage changeSettingValue(String settingKey, int index, String value) {
    Selenide.$("[data-key=\"" + settingKey + "\"]").findAll("input").get(index).val(value);
    return this;
  }

  public SettingsPage clickOnCancel(String settingKey) {
    click(Selenide.$("[data-key=\"" + settingKey + "\"]").find(byText("Cancel")));
    return this;
  }

  public SettingsPage removeFirstValue(String settingKey) {
    click(Selenide.$("[data-key=\"" + settingKey + "\"]").find(".js-remove-value"));
    return this;
  }

  public SettingsPage sendDeleteKeyToSettingField(String settingKey) {
    Selenide.$("[data-key=\"" + settingKey + "\"]").find("input").sendKeys(Keys.BACK_SPACE);
    return this;
  }

  public SettingsPage setStringValue(String settingKey, String value) {
    SelenideElement setting = Selenide.$(".settings-definition[data-key=\"" + settingKey + "\"]");
    setting.find("input").val(value);
    setting.find(".button-success").click();
    setting.find(".button-success").shouldNot(Condition.exist);
    return this;
  }

  public PropertySetInput getPropertySetInput(String settingKey) {
    SelenideElement setting = Selenide.$(".settings-definition[data-key=\"" + settingKey + "\"]");
    return new PropertySetInput(setting);
  }

  private void click(SelenideElement selenideElement){
    // FIXME Temporary fix to correctly scroll in Firefox 46
    Selenide.executeJavaScript("arguments[0].click()", selenideElement);
  }

}

File: server/sonar-web/src/main/js/apps/settings/store/settingsPage.ts
/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { omit } from 'lodash';
import { combineReducers } from 'redux';
import { ActionType } from '../../../store/utils/actions';

const enum Actions {
  CancelChange = 'settingsPage/CANCEL_CHANGE',
  ChangeValue = 'settingsPage/CHANGE_VALUE',
  FailValidation = 'settingsPage/FAIL_VALIDATION',
  PassValidation = 'settingsPage/PASS_VALIDATION',
  StartLoading = 'settingsPage/START_LOADING',
  StopLoading = 'settingsPage/STOP_LOADING'
}

type Action =
  | ActionType<typeof cancelChange, Actions.CancelChange>
  | ActionType<typeof changeValue, Actions.ChangeValue>
  | ActionType<typeof failValidation, Actions.FailValidation>
  | ActionType<typeof passValidation, Actions.PassValidation>
  | ActionType<typeof startLoading, Actions.StartLoading>
  | ActionType<typeof stopLoading, Actions.StopLoading>;

export interface State {
  changedValues: T.Dict<any>;
  loading: T.Dict<boolean>;
  validationMessages: T.Dict<string>;
}

export function cancelChange(key: string) {
  return { type: Actions.CancelChange, key };
}

export function changeValue(key: string, value: any) {
  return { type: Actions.ChangeValue, key, value };
}

function changedValues(state: State['changedValues'] = {}, action: Action) {
  if (action.type === Actions.ChangeValue) {
    return { ...state, [action.key]: action.value };
  }
  if (action.type === Actions.CancelChange) {
    return omit(state, action.key);
  }
  return state;
}

export function failValidation(key: string, message: string) {
  return { type: Actions.FailValidation, key, message };
}

export function passValidation(key: string) {
  return { type: Actions.PassValidation, key };
}

function validationMessages(state: State['validationMessages'] = {}, action: Action) {
  if (action.type === Actions.FailValidation) {
    return { ...state, [action.key]: action.message };
  }
  if (action.type === Actions.PassValidation) {
    return omit(state, action.key);
  }
  return state;
}

export function startLoading(key: string) {
  return { type: Actions.StartLoading, key };
}

export function stopLoading(key: string) {
  return { type: Actions.StopLoading, key };
}

function loading(state: State['loading'] = {}, action: Action) {
  if (action.type === Actions.StartLoading) {
    return { ...state, [action.key]: true };
  }
  if (action.type === Actions.StopLoading) {
    return { ...state, [action.key]: false };
  }
  return state;
}

export default combineReducers({ changedValues, loading, validationMessages });

export function getChangedValue(state: State, key: string) {
  return state.changedValues[key];
}

export function getValidationMessage(state: State, key: string): string | undefined {
  return state.validationMessages[key];
}

export function isLoading(state: State, key: string) {
  return Boolean(state.loading[key]);
}
